Use flatware to parallelize CI specs

mjankowski opened the pull request:
In addition to https://github.com/mastodon/mastodon/pull/27660 - here's another idea for maintaining some parallelization setup, while not paying the cost of all the setup times.

Opening this initially because I want to see if the CI setup/speedup is as easy as it was locally. Will review the actual numbers and post about that after it completes...

mjankowski committed:
Use flatware to parallelize CI specs

mjankowski commented:
It's challening to get good numbers here because it's so variable on each run. Some recent actions runs (all from after merging in the matrix reduction):

|PR|3.0|3.1|3.2|
|-|-|-|-|
|27660|~9m|~12m|~8:30|
|26559|~14:30|11m|~9m|
|8214|9|12|8|
|Flatware PR|4|7|7|

Those are all looking just at the narrow rspec run portion of the larger job. I think the discrepancy in the flatware times is that the ruby 3.0 job is using 4 cores and the others are using 2 each.

Another benefit here might be to cache the rspec status persistence file between runs -- if it's present flatware can use it to split tests between cores based on exected times (it defaults to just a numeric split w/out that file).

I think this probably makes sense to apply, but I can also wait until we have more runtime data post-matrix-removal and see what things actually look like.

nschonni commented on Gemfile:
```diff
@@ -198,3 +198,5 @@ gem 'net-http', '~> 0.3.2'
 
 gem 'hcaptcha', '~> 7.1'
+
+gem 'flatware-rspec', '~> 2.1'
```
Probably should be up in the `:development` section where the `rspec-chunked` was

mjankowski replied:
Good catch - I had `bundle add`'d it during testing and never moved into group. Done.

mjankowski commented:
Rebased again. These run times remain variable due to differences in underlying hardware/env, but recently the full runs of "Ruby Testing" task for successful runs is in the ~13-15 minute zone. This PRs most recent rebased run was just over ~8 mins.

The narrow rspec run portion of that (for each ruby version) tends to be in the 8-12 min zone (again, quite variable) on the non-parallel runs. This PRs most recent was ~4 mins. I suspect part of the variability is in how many cores happen to be allocated for a given runner ... so ~15 mins might be a worst-case on the non-parallel runs and just under ~4 mins might be expected on parallel runs when the runner gets 4 cores.

Maybe as importantly - none of these runs have shown any intermittent failures or other parallelism issues; so I think this is probably safe if we want a CI speedup.

mjankowski commented:
With the sidekiq fake PR merged, this one is probably the next most impactful (in terms of CI performance).

Rebased this again to include the sidekiq improvements here.

When the flatware runners get 4 cores on CI, looks like the most recent runs were just over ~3 mins each for the full rspec run portion of the CI suite (vs 6-7 mins for other recent post-sidekiq-merge runs from other PRs).

mjankowski commented:
Rebased again after we added ruby 3.3 to the CI matrix.

It continues to vary, but recent runs have been in the ~5:45-6:30 or so range for the rspec portion. Notably, the 3.3 rspec runs are generally on the shorter end - just under 6 mins (very small sample size in this observation).

The recent runs for this PR are in the ~3min range, with the ruby 3.3 runs all under 3 mins.

I'd love for literally anyone to comment on this PR, because if we are not going to merge it seems unproductive to keep talking to myself about it and rebasing.

renchap commented:
It cuts CI times in half so it is probably a good idea, even if it introduces another layer in our test framework. As the changes are not large, it would also be easy to revert if we discover it causes issues.

@ClearlyClaire what do you think?

ClearlyClaire commented:
One of the tests failed with:
> A worker died!

I don't know if we can get more details. Also, not sure if that makes tests less reliable or if this is negligible.

mjankowski commented:
Might be related to regression in latest version - https://github.com/briandunn/flatware/issues/84 (I recently rebased and version bumped within this PR).

mjankowski commented:
Rebased and added two changes:

- Revert previous version bump here which was causing the worker died issues -- will watch the linked issue and update again when resolved
- Added a flatware<>simplecov integration config which repairs the simplecov reports by combining coverage results from the parallelized runs at the end.

mjankowski commented:
There was a configuration interaction issue here briefly ... since simplecov is only loaded/run when the ruby matrix version matches ruby-version on CI, the other versions (currently 3.1 and 3.3) were failing/timing-out their CI runs. Pushed an update to only set up the flatware fork config when we want simplecov active. I think that will resolve it ... but will monitor and publish another status summary after this CI run.

mjankowski commented:
Looks like that last push resolved the CI issues, and we are now re-combining simplecov output locally correctly.

Looks like codecov is not combining though, and needs more configuration to do that. Will attempt that next.

mjankowski commented:
The codecov integration is the last remaining portion of this to sort out.

The first few obvious seeming things did not work. Before I spend more time on that ... is the codecov integration useful? I rarely look at it, but when I do it seems confusing. Like right now its comparing recent open PRs to a commit from ~3 months ago? I don't know if the service only sort of works, or if we broke the integration somehow, or if I'm interpreting it wrong?

briandunn commented:
@mjankowski simplecov support is a pretty new flatware feature. If that's a blocker here I'd love to help. Would be cool to see this merge.

mjankowski commented:
@briandunn thanks for the offer (and for flatware, it's great!). I think at this point...

- The basic integration here was easy to do and is all set up and working correctly. We see expected speedup both locally and on CI.
- The interaction between rspec/flatware/simplecov is set up according to your integration docs, and I believe working correctly - results from each parallel run seem to be combined and report the full coverage
- The part of the CI process which uploads the coverage result data to the hosted codecov service is not fully updated yet ... because a) I'm not 100% sure on what needs to change to make it work, b) I'm waiting for an answer on whether anyone actually looks at this integration (and want to just disable it if no one cares), c) the codecov/GH feedback loop is pretty slow and sort of confusing and thus I have not spent a ton of time on it.

I found one load order oddity (not a bug, just mistake I made) when I was working on how to configure the flatware/simplecov integration ... maybe I'll open an issue or documentation update for you on flatware repo.

nschonni commented on bin/flatware:
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+#
+# This file was generated by Bundler.
```
If this is generated when running bundler, should it be gitignored?

mjankowski replied:
No, it's an intentionally added [bundler binstub](https://bundler.io/v2.5/man/bundle-binstubs.1.html)
